tools/flash_shelly.py

```diff
@@ -263,6 +263,7 @@ def write_flash(host, lfw, dlurl, cfw_type, mode):
       checkurl = f'http://{host}/rpc/Shelly.GetInfo'
     else:
       checkurl = f'http://{host}/Shelly.GetInfo'
+      time.sleep(17) # some stock devices require time to complete flash process.
```

Comment:
why 17? and why not increase the number of attempts we allow in the condition above?

Reply:
it not that, it the devcie stays online whilst it is flashing, so it does not get wait looped, and it can take 15 sec beofre it goes offline to reboot.

Reply:
your fw seems it shut down rpc whilst it's flashing, so it reports error and thus it gets wait looped.

Reply:
so unless you can think of a way we can detect if a flash in progress?

Reply:
see what i mean here
```
Do you wish to flash shellydimmer-F3CD89 to firmware version 1.8.4 (y/n) ? y

write_flash
Now Flashing...
curl -qsS http://shellydimmer-F3CD89.local/ota?url=http://archive.shelly-faq.de/version/v1.8.4/SHDM-1.zip
{"status":"updating","has_update":false,"new_version":"20201007-085347/v1.8.5@5fbcbaa3","old_version":"20201007-085347/v1.8.5@5fbcbaa3"}
waiting for shellydimmer-F3CD89 to reboot
INFO: {'type': 'SHDM-1', 'mac': 'CC50E3F3CD89', 'auth': False, 'fw': '20201007-085347/v1.8.5@5fbcbaa3', 'num_outputs': 1, 'num_meters': 1}
INFO: {'type': 'SHDM-1', 'mac': 'CC50E3F3CD89', 'auth': False, 'fw': '20201007-085347/v1.8.5@5fbcbaa3', 'num_outputs': 1, 'num_meters': 1}
INFO: {'type': 'SHDM-1', 'mac': 'CC50E3F3CD89', 'auth': False, 'fw': '20201007-085347/v1.8.5@5fbcbaa3', 'num_outputs': 1, 'num_meters': 1}
INFO: {'type': 'SHDM-1', 'mac': 'CC50E3F3CD89', 'auth': False, 'fw': '20201007-085347/v1.8.5@5fbcbaa3', 'num_outputs': 1, 'num_meters': 1}
INFO: {'type': 'SHDM-1', 'mac': 'CC50E3F3CD89', 'auth': False, 'fw': '20201007-085347/v1.8.5@5fbcbaa3', 'num_outputs': 1, 'num_meters': 1}
INFO: {'type': 'SHDM-1', 'mac': 'CC50E3F3CD89', 'auth': False, 'fw': '20201007-085347/v1.8.5@5fbcbaa3', 'num_outputs': 1, 'num_meters': 1}
INFO: {'type': 'SHDM-1', 'mac': 'CC50E3F3CD89', 'auth': False, 'fw': '20201007-085347/v1.8.5@5fbcbaa3', 'num_outputs': 1, 'num_meters': 1}
INFO: {'type': 'SHDM-1', 'mac': 'CC50E3F3CD89', 'auth': False, 'fw': '20201007-085347/v1.8.5@5fbcbaa3', 'num_outputs': 1, 'num_meters': 1}
INFO: {'type': 'SHDM-1', 'mac': 'CC50E3F3CD89', 'auth': False, 'fw': '20201007-085347/v1.8.5@5fbcbaa3', 'num_outputs': 1, 'num_meters': 1}
INFO: {'type': 'SHDM-1', 'mac': 'CC50E3F3CD89', 'auth': False, 'fw': '20201007-085347/v1.8.5@5fbcbaa3', 'num_outputs': 1, 'num_meters': 1}
INFO: {'type': 'SHDM-1', 'mac': 'CC50E3F3CD89', 'auth': False, 'fw': '20201007-085347/v1.8.5@5fbcbaa3', 'num_outputs': 1, 'num_meters': 1}
INFO: {'type': 'SHDM-1', 'mac': 'CC50E3F3CD89', 'auth': False, 'fw': '20201007-085347/v1.8.5@5fbcbaa3', 'num_outputs': 1, 'num_meters': 1}
INFO: {'type': 'SHDM-1', 'mac': 'CC50E3F3CD89', 'auth': False, 'fw': '20201007-085347/v1.8.5@5fbcbaa3', 'num_outputs': 1, 'num_meters': 1}
INFO: {'type': 'SHDM-1', 'mac': 'CC50E3F3CD89', 'auth': False, 'fw': '20201007-085347/v1.8.5@5fbcbaa3', 'num_outputs': 1, 'num_meters': 1}
INFO: {'type': 'SHDM-1', 'mac': 'CC50E3F3CD89', 'auth': False, 'fw': '20201007-085347/v1.8.5@5fbcbaa3', 'num_outputs': 1, 'num_meters': 1}
INFO: {'type': 'SHDM-1', 'mac': 'CC50E3F3CD89', 'auth': False, 'fw': '20201007-085347/v1.8.5@5fbcbaa3', 'num_outputs': 1, 'num_meters': 1}
INFO: {'type': 'SHDM-1', 'mac': 'CC50E3F3CD89', 'auth': False, 'fw': '20200904-094531/v1.8.4@699b08ac', 'num_outputs': 1, 'num_meters': 1}
INFO: {'type': 'SHDM-1', 'mac': 'CC50E3F3CD89', 'auth': False, 'fw': '20200904-094531/v1.8.4@699b08ac', 'num_outputs': 1, 'num_meters': 1}
INFO: {'type': 'SHDM-1', 'mac': 'CC50E3F3CD89', 'auth': False, 'fw': '20200904-094531/v1.8.4@699b08ac', 'num_outputs': 1, 'num_meters': 1}
INFO: {'type': 'SHDM-1', 'mac': 'CC50E3F3CD89', 'auth': False, 'fw': '20200904-094531/v1.8.4@699b08ac', 'num_outputs': 1, 'num_meters': 1}
```

Reply:
shelly-homekit stops HAP service while firmware is updating but still responds to the GetInfo rpc.

i suggest the following: after starting flashing, you start a loop with successful exit condition being version == expected version, or at least version != old version. while looping, you ignore url errors (as we expect the device to reboot at some point) as well as successful responses where version == old_version (update ongoing).

Reply:
and just timeout after how long, for a real failure ?

Reply:
i'd make it 15 attempts with 2 second sleep in between.

Reply:
ok.

Reply:
ok done.

Reply:
oops sec still on 41

Reply:
ok now done